### ISSUE TITLE
Semantics of no-body functions

### ISSUE THREAD

**Issue by fbrausse** [esbmc/esbmc#604]
Consider this example:
```c
typedef int func();

func *fun();

int main()
{
	func *f = fun();
	int x = f ? f() : 0;
	assert(!x);
}
```
ESBMC detects
```
Counterexample:

State 1 file code6.c line 9 function main thread 0
----------------------------------------------------
  f = INVALID2

State 3 file code6.c line 10 function main thread 0
----------------------------------------------------
  x = 1 (00000000 00000000 00000000 00000001)

State 4 file code6.c line 11 function main thread 0
----------------------------------------------------
Violated property:
  file code6.c line 11 function main
  assertion
  !(_Bool)x


VERIFICATION FAILED
```
while CBMC (even with the `--havoc-undefined-functions` flag) says `VERIFICATION SUCCESSFUL`.

It seems semantic interpretation differs. Why? Am I using the tools wrongly; are there flags to bring them into alignment?

**Comment by rafaelsamenezes** [comment]
I am not sure I understood the question. In summary, we consider a function without a body as nondeterministic. In your example: 

1. You declare a function that returns a function pointer.
2. You initialize a pointer calling it. Since the `fun` function is nondet, it will generate an invalid object.
3. Dereferencing an invalid-object as a function will cause another nondet behaviour.

**Comment by fbrausse** [comment]
The question is about the different verdicts given by ESBMC and CBMC for the example and whether there are any flags to bring them to the same semantics - both ways: `cbmc --pointer-check` results in a negative verdict (same as ESBMC), but there does not seem to be a way to make ESBMC accept the code - which looks dubious, given that ESBMC has no way of proving an actual bug there: `INVALID2` cannot be constructed by the program - ever - and thus there is no way to prove that a call to `INVALID2` results in `x=1`.

**Comment by fbrausse** [comment]
In light of the UB checker missing, it may even be argued that by standard C there is no way to construct a function pointer pointing anything else than a function (symbol) defined in the final program: casting `void *` to an fptr or vice versa is UB. The only way to construct an fptr is by referencing an existing function. ESBMC here does not see the "final" program and thus cannot assume that `fun()` actually uses UB to construct an invalid object.

**Comment by mikhailramalho** [comment]
> The question is about the different verdicts given by ESBMC and CBMC for the example and whether there are any flags to bring them to the same semantics - both ways: `cbmc --pointer-check` results in a negative verdict (same as ESBMC), but there does not seem to be a way to make ESBMC accept the code - which looks dubious, given that ESBMC has no way of proving an actual bug there: `INVALID2` cannot be constructed by the program - ever - and thus there is no way to prove that a call to `INVALID2` results in `x=1`.

I think the point here is that `INVALID2` can be anything, so esbmc checks both branches of the ternary operator. Maybe a better name would be `UNKNOWN`? 

I don't remember the details, but I thought invalid pointers were only constructed when there is (potentially) bad dereference.

**Comment by fbrausse** [comment]
I agree that `INVALID2` is the main point of contention here. But not just the naming: Appearently ESBMC assumes `fun()` to return a pointer to something that actually can be called, but then assumes the return value can be anything. To me this looks  a bit arbitrary in the sense that the assumption that `INVALID2` can actually be executed is in contradiction with the assumption that whatever it returns has no assumptions made on - and can therefore violate the `assert(!x)`. Why is `INVALID2` assumed to be callable in the first place then?
